### DISCUSSION TITLE
Dropping globals

### DISCUSSION THREAD

**Discussion by filasieno** [sunbains/embedded-innodb#65] [Ideas]
Hi,

This is the direction/strategy that I would propose to drop the globals; I think it can be done very quickly.

# Global Drop Plan

### Step 1: Global Variable Consolidation

Move all **extern globals**, **file-level static variables**, and **class-level static variables** to a single state variable.
Globals are sprinkled all around. We need to consolidate and reclassify.

This is just a mechanical step.

Something like:

```C++
class InnoDB_state {
   // Server module
   bool srv_xxx;
   bool srv_yyy;

   // OS module
   bool os_xxx;
   // Fil module
   // Btr module 
   ...
}
```

This is how it would look: [InnoDB_state](https://github.com/filasieno/embedded-innodb/blob/drop-globals/include/srv0state.h#L147)

### Step 2: Pass the State Variable as Argument

Update code to accept the state variable as an argument or create classes that have a single pointer to the state variable to implement the key components.
This way the code is re-entrant.

### Step 3: Reorganize Source Code

- Check all includes
- Check all dependencies
- Minimize includes in header files (to improve compile times)
- Clearly classify headers designed to be published as API from headers used internally (John Lakos style)
- Define clear source code guidelines
- Add static analysis

### Step 4: Test lab setup

Add/Improve Unit Tests, Code Coverage, Integration Tests, Deterministic Testing Setup, Performance Testing Setup, Tracing Setup, Tooling Setup (Visualization Tools).

Basically, set up a test lab and then test, test, test, test, test, ...

# Current State of Tests

Tests are failing; therefore, the globals refactoring would be best effort.
Is there a clear set of tests that I can use for regression testing for the `global drop` refactoring?

# Questions

1. Is this direction something we can agree on? Should I go forward (it can be done quickly)? If no, what should we do? I'm tentatively committed to the project; I'd like to see if it moves forward.
2. Any ideas for regression checks during the global refactoring?
3. Another friend (who is a good build engineer) can help too; but... What is the target: Zig or C++? If it is Zig (which I prefer), moving to C++ doesn't seem very wise. C++ is quite complex and tricky; if the target is "Modern C++," for me it is a show-stopper as I do not enjoy it.

**Let me know how I can help.**

**Comment by sunbains** [comment]
1. What you propose is the last step in my current plan. My preference is to first understand all the dependencies fully. The reason is that I want to compose the classes at will. Eg., single IO instance shared by different components or multiple buffer pools that share a single page flush instance, DD instance per schema, how will FKs work if we have say DD per Schema? etc. Multiple buffer pool instances, eg. one per block size, it would be required for implementing an LSM store.

2. Zig (not decided yet, if it happens is the last step), I very much like modern C++.

3. I personally don't see any particular value in Zig over modern C++. My motivation is a bit different, BTW the rewrite could be in Rust too. My motivation is to rewrite it under a more permissive license like MIT/BSD etc. choosing a different language is to avoid any lingering copyright issues.

4. By refactoring into classes, I think the port to another language will be a lot easier. By implementing the leaf classes in Zig/Rust.

**Reply by filasieno** [reply]
# Database Architecture: Component Dependencies and Design Approaches

I completely agree with your emphasis on understanding dependencies first. Your concerns about component composition are crucial:

- Shared I/O instance requirements
- Multiple buffer pools with shared page flush
- Per-schema data dictionary instances
- Foreign key handling across schemas
- LSM store support considerations

These are exactly the kinds of architectural challenges that I find interesting to address.

However, I see some significant risks with the class-based refactoring approach:

## Complex Component Interactions
Consider a seemingly simple operation like page eviction:
- Buffer manager needs lock manager state
- Might require transaction manager interaction
- Could need memory manager coordination
- May involve I/O manager for flush
- Potentially affects multiple buffer pools
- Any of these interactions might require task suspension

These aren't just implementation details—they're fundamental requirements. Class encapsulation tends to hide these real dependencies behind interfaces that don't match reality or, by definition, need to change as often as new techniques need to be introduced.

## Threading and Task Complexity
In a database environment:
- Tasks can block at any point, requiring efficient suspension/resumption
- State transitions must be trackable
- Resource ownership is dynamic
- Deadlock prevention requires global awareness

Traditional class boundaries make these harder to manage because they enforce artificial encapsulation where we actually need transparency.

While I see the value in abstracting via classes, it is important to bear in mind the critical issues that could arise in terms of performance—a point to carefully evaluate and manage.

Another approach, which would share the initial steps you have proposed, is the **Data-Oriented State-Centric Design**.

## Structured State Organization
```
DatabaseState:
  MemoryState:
    BufferPools[]    // Multiple instances possible, shared flushers
    Allocators       // Global or per-component
  ConcurrencyState:
    LockTables       // Cross-component coordination
    TransactionState // Global visibility needed
  StorageState:
    FileState
    LogState
  SchemaState:
    DDictionaries[]  // Per schema
    FKManagement
```

## Key Principles
- State transitions are explicit and trackable
- Components can access required state directly 
- Data access patterns are explicit and measured
- Natural support for task suspension/resumption

This approach has proven successful in other complex domains (game engines, high-frequency trading systems) that share similar characteristics:
- Need for predictable performance
- Complex state management
- Heavy concurrency requirements
- Task-based execution models

## Language Choice Considerations
The case for Zig is that its ecosystem puts significant emphasis on data-oriented design principles.

### Data-First Philosophy
- Explicit allocator management
- Focus on data structure layout
- Structure-of-Arrays vs. Array-of-Structures ready-made patterns
- Index-based entity management rather than pointer-heavy designs

### Modern C++ vs. Zig Approaches
- C++ community often focuses heavily on code abstraction idioms
- This leads to emphasis on behavior rather than data organization
- Zig community emphasizes data organization first
- Memory layout and allocation patterns are explicit and central
- Better alignment with database system needs where data organization is crucial

### Practical Benefits for Databases
- Easier implementation of buffer pool layouts
- Natural expression of page management
- Better control over memory patterns
- Simpler implementation of task suspension points
- More predictable performance characteristics

While Modern C++ is powerful, its complexity often pushes developers toward abstraction-heavy designs that can obscure the fundamental data management needs of a database system.
This actually reinforces the case for a data-oriented, state-centric design approach regardless of final language choice. However, Zig's ecosystem would make it more natural to implement such designs.

**Note:** A state-centric/data-oriented approach can be implemented in any low-level language, and this is the part I'm concerned with.

## Next Steps
The first step—collecting globals into a structured state—is valuable for both approaches. It gives us visibility into the real dependencies and state transitions.

Would love to hear your thoughts on this synthesis.

**Should I move the globals into a single variable as proposed and then see how it goes from there?**

**Comment by sunbains** [comment]
1. I don't really believe that you can't do a data-centric design in C++. You can do what ever you want. So, I don't buy that.
2. The InnoDB code is already laid out as classes, just happen to be done in a C way. Therefore it's very easy to change to classes.
3. InnoDB function API across components works fine, it's quite transparent. The problem is the bit rot over the years to make it work with MySQL.
4. The layout of the buffer pool is already optimal, the memory is allocated in chunks which are then divvyed into pages. These are fixed at startup. The memory allocation in this version of InnoDB relies very little on new/delete. The high frequency small objets are the logical locks and their backing store is the buffer pool memory. It's quite well thought out already. So that's not something that worries me.
5. The abstractions for task suspension points is a far more complicated problem but the abstraction for that already exists in this design. It was never finished, that's all.  It's the que_thr_t struct. Underneath the hood is a simple virtual machine. Understanding all these internal mechanisms is more important.
6. The state transitions are already explicit for the que_thr_t states.
7. Passing a global instance around as a data member for access is really a trivial part of the bigger problem and I don't think there is any disagreement on that. I just don't think it's possible to do without first converging all the components into classes so that we can add the data members to them.

**Reply by filasieno** [reply]
I agree on everything especially point 1. 
```
A state-centric/data-oriented approach can be implemented in any low-level language, and this is the part I'm concerned with.
```

All the different points I think that they are easier with simple access to state.

Shall I move the globals then?

**Comment by sunbains** [comment]
If you mean putting them all in a class called InnoDB then sure. There will probably be lots of conflicts while I make everything into a class.

**Reply by filasieno** [reply]
Ok.

I'll put all globals in a class InnoDB_state; then you can rename reorganize as you will, just because the InnoDB class is already there.  I'll carefully do the merge incrementally as you push changes.

**Reply by sunbains** [reply]
Just call it InnoDB. State is implied.

**Reply by filasieno** [reply]
Ok
